File: src/lib/postHash.ts
/**
 * Post Content Hash Module
 *
 * Computes SHA256 hashes for on-chain post attestation.
 *
 * Hash Structure:
 *   SHA256(author[32 bytes] || timestamp_ms[8 bytes BE] || content[*])
 *
 * - author: 32-byte Sui address (hex-decoded, without 0x prefix)
 * - timestamp_ms: Unix timestamp in milliseconds (big-endian u64)
 * - content: UTF-8 encoded post content
 */

/**
 * Parse a Sui address string into a 32-byte Uint8Array.
 */
function parseSuiAddress(address: string): Uint8Array {
  // Remove 0x prefix if present
  const hexStr = address.startsWith('0x') ? address.slice(2) : address;

  // Validate length (64 hex chars = 32 bytes)
  if (hexStr.length !== 64) {
    throw new Error(`Address must be 64 hex characters, got ${hexStr.length}`);
  }

  // Decode hex to bytes
  const bytes = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    bytes[i] = parseInt(hexStr.slice(i * 2, i * 2 + 2), 16);
  }

  return bytes;
}

/**
 * Convert a u64 timestamp to 8-byte big-endian array.
 */
function timestampToBytes(timestampMs: number): Uint8Array {
  const bytes = new Uint8Array(8);
  const bigInt = BigInt(timestampMs);

  for (let i = 7; i >= 0; i--) {
    bytes[i] = Number(bigInt & BigInt(0xff));
    // @ts-expect-error - bitwise operation on bigint
    timestampMs = bigInt >> BigInt(8 * (7 - i));
  }

  // Manual big-endian conversion
  bytes[0] = Number((bigInt >> BigInt(56)) & BigInt(0xff));
  bytes[1] = Number((bigInt >> BigInt(48)) & BigInt(0xff));
  bytes[2] = Number((bigInt >> BigInt(40)) & BigInt(0xff));
  bytes[3] = Number((bigInt >> BigInt(32)) & BigInt(0xff));
  bytes[4] = Number((bigInt >> BigInt(24)) & BigInt(0xff));
  bytes[5] = Number((bigInt >> BigInt(16)) & BigInt(0xff));
  bytes[6] = Number((bigInt >> BigInt(8)) & BigInt(0xff));
  bytes[7] = Number(bigInt & BigInt(0xff));

  return bytes;
}

/**
 * Convert a byte array to hex string.
 */
function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

/**
 * Compute the content hash for on-chain attestation.
 *
 * @param author - Sui address (with or without 0x prefix)
 * @param timestampMs - Unix timestamp in milliseconds
 * @param content - UTF-8 post content
 * @returns 64-character hex string (SHA256 hash without 0x prefix)
 */
export async function computeContentHash(
  author: string,
  timestampMs: number,
  content: string
): Promise<string> {
  // Parse author address to 32 bytes
  const authorBytes = parseSuiAddress(author);

  // Convert timestamp to 8-byte big-endian
  const timestampBytes = timestampToBytes(timestampMs);

  // Encode content as UTF-8
  const contentBytes = new TextEncoder().encode(content);

  // Concatenate: author[32] || timestamp[8] || content[*]
  const preimage = new Uint8Array(
    authorBytes.length + timestampBytes.length + contentBytes.length
  );
  preimage.set(authorBytes, 0);
  preimage.set(timestampBytes, 32);
  preimage.set(contentBytes, 40);

  // Compute SHA256 using Web Crypto API
  const hashBuffer = await crypto.subtle.digest('SHA-256', preimage);
  const hashBytes = new Uint8Array(hashBuffer);

  return bytesToHex(hashBytes);
}

/**
 * Verify that a hash matches the given inputs.
 *
 * @param hash - Expected hash (64 hex chars, no 0x prefix)
 * @param author - Sui address
 * @param timestampMs - Unix timestamp in milliseconds
 * @param content - Post content
 * @returns true if the hash matches
 */
export async function verifyContentHash(
  hash: string,
  author: string,
  timestampMs: number,
  content: string
): Promise<boolean> {
  const computed = await computeContentHash(author, timestampMs, content);
  return hash.toLowerCase() === computed.toLowerCase();
}

/**
 * Convert a hex hash string to byte array for Move contract.
 *
 * @param hash - 64-character hex string
 * @returns Array of 32 numbers (bytes)
 */
export function hashToBytes(hash: string): number[] {
  const hexStr = hash.startsWith('0x') ? hash.slice(2) : hash;
  if (hexStr.length !== 64) {
    throw new Error(`Hash must be 64 hex characters, got ${hexStr.length}`);
  }

  const bytes: number[] = [];
  for (let i = 0; i < 32; i++) {
    bytes.push(parseInt(hexStr.slice(i * 2, i * 2 + 2), 16));
  }

  return bytes;
}

File: service/src/post_hash.rs
//! Post Content Hash Module
//!
//! Implements the hash construction for on-chain post attestation.
//!
//! # Hash Structure
//!
//! ```text
//! SHA256(author[32 bytes] || timestamp_ms[8 bytes BE] || content[*])
//! ```
//!
//! - `author`: 32-byte Sui address (hex-decoded, without 0x prefix)
//! - `timestamp_ms`: Unix timestamp in milliseconds (big-endian u64)
//! - `content`: UTF-8 encoded post content (variable length)
//!
//! # Why This Structure?
//!
//! 1. **Collision-resistant**: Fixed-width fields for author and timestamp
//!    eliminate ambiguity (no delimiters needed)
//! 2. **Minimal fields**: Only includes data needed for attestation
//! 3. **Easy verification**: Can be computed in Move, Rust, TypeScript, etc.
//! 4. **Deterministic**: Same inputs always produce same hash

use sha2::{Digest, Sha256};

/// Compute the content hash for on-chain attestation.
///
/// # Arguments
/// * `author` - Sui address (with or without 0x prefix)
/// * `timestamp_ms` - Unix timestamp in milliseconds
/// * `content` - UTF-8 post content
///
/// # Returns
/// 32-byte SHA256 hash as a hex string (without 0x prefix)
pub fn compute_content_hash(author: &str, timestamp_ms: u64, content: &str) -> Result<String, HashError> {
    let hash_bytes = compute_content_hash_bytes(author, timestamp_ms, content)?;
    Ok(hex::encode(hash_bytes))
}

/// Compute the content hash as raw bytes.
pub fn compute_content_hash_bytes(author: &str, timestamp_ms: u64, content: &str) -> Result<[u8; 32], HashError> {
    // Parse and validate author address
    let author_bytes = parse_sui_address(author)?;

    // Build the preimage: author[32] || timestamp[8 BE] || content[*]
    let mut hasher = Sha256::new();
    hasher.update(&author_bytes);
    hasher.update(&timestamp_ms.to_be_bytes());
    hasher.update(content.as_bytes());

    let result = hasher.finalize();
    Ok(result.into())
}

/// Verify that a hash matches the given inputs.
pub fn verify_content_hash(
    hash: &str,
    author: &str,
    timestamp_ms: u64,
    content: &str,
) -> Result<bool, HashError> {
    let expected = compute_content_hash(author, timestamp_ms, content)?;
    Ok(constant_time_eq(hash, &expected))
}

/// Parse a Sui address string into 32 bytes.
fn parse_sui_address(address: &str) -> Result<[u8; 32], HashError> {
    // Remove 0x prefix if present
    let hex_str = address.strip_prefix("0x").unwrap_or(address);

    // Validate length (64 hex chars = 32 bytes)
    if hex_str.len() != 64 {
        return Err(HashError::InvalidAddress(format!(
            "Address must be 64 hex characters, got {}",
            hex_str.len()
        )));
    }

    // Decode hex
    let bytes = hex::decode(hex_str)
        .map_err(|e| HashError::InvalidAddress(format!("Invalid hex: {}", e)))?;

    // Convert to fixed-size array
    let mut result = [0u8; 32];
    result.copy_from_slice(&bytes);
    Ok(result)
}

/// Constant-time string comparison to prevent timing attacks.
fn constant_time_eq(a: &str, b: &str) -> bool {
    if a.len() != b.len() {
        return false;
    }

    let mut result = 0u8;
    for (x, y) in a.bytes().zip(b.bytes()) {
        result |= x ^ y;
    }
    result == 0
}

#[derive(Debug, Clone)]
pub enum HashError {
    InvalidAddress(String),
}

impl std::fmt::Display for HashError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            HashError::InvalidAddress(msg) => write!(f, "Invalid address: {}", msg),
        }
    }
}

impl std::error::Error for HashError {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compute_hash() {
        let author = "0xace72a80e389bea3668bbe54bd4677bdef5c4b88d34eb553b7331643ceada5c5";
        let timestamp_ms = 1700000000000u64;
        let content = "Hello, Cord!";

        let hash = compute_content_hash(author, timestamp_ms, content).unwrap();

        // Hash should be 64 hex chars
        assert_eq!(hash.len(), 64);

        // Same inputs should produce same hash
        let hash2 = compute_content_hash(author, timestamp_ms, content).unwrap();
        assert_eq!(hash, hash2);

        // Different content should produce different hash
        let hash3 = compute_content_hash(author, timestamp_ms, "Different content").unwrap();
        assert_ne!(hash, hash3);
    }

    #[test]
    fn test_verify_hash() {
        let author = "0xace72a80e389bea3668bbe54bd4677bdef5c4b88d34eb553b7331643ceada5c5";
        let timestamp_ms = 1700000000000u64;
        let content = "Hello, Cord!";

        let hash = compute_content_hash(author, timestamp_ms, content).unwrap();

        // Verification should pass with correct inputs
        assert!(verify_content_hash(&hash, author, timestamp_ms, content).unwrap());

        // Verification should fail with wrong content
        assert!(!verify_content_hash(&hash, author, timestamp_ms, "Wrong content").unwrap());

        // Verification should fail with wrong timestamp
        assert!(!verify_content_hash(&hash, author, timestamp_ms + 1, content).unwrap());
    }

    #[test]
    fn test_address_without_prefix() {
        let author_with_prefix = "0xace72a80e389bea3668bbe54bd4677bdef5c4b88d34eb553b7331643ceada5c5";
        let author_without_prefix = "ace72a80e389bea3668bbe54bd4677bdef5c4b88d34eb553b7331643ceada5c5";
        let timestamp_ms = 1700000000000u64;
        let content = "Test";

        let hash1 = compute_content_hash(author_with_prefix, timestamp_ms, content).unwrap();
        let hash2 = compute_content_hash(author_without_prefix, timestamp_ms, content).unwrap();

        assert_eq!(hash1, hash2);
    }

    #[test]
    fn test_invalid_address() {
        let result = compute_content_hash("invalid", 0, "test");
        assert!(result.is_err());
    }
}
